util/genrest/gomodel/pathtemplateparser.go

```diff
@@ -0,0 +1,265 @@
+		seg, err := parse()
+		if err != nil || seg != nil {
+			return seg, err
+		}
+	}
+	return nil, fmt.Errorf("could not parse path segment")
+}
+
+// parseSingleValue parses a segment with `Kind==SingleValue` (i.e. a single-segment placeholder), returning nil if not possible.
+func (parser *Parser) parseSingleValue() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reSingleValue, `^\*`)
+	return parser.parseToSegment(re, SingleValue)
+}
+
+// parseMultipleValue parses a segment with `Kind==MultipleValue` (i.e. a multiple-segment placeholder), returning nil if not possible.
+func (parser *Parser) parseMultipleValue() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reMultipleValue, `^\*\*`)
+	seg, err := parser.parseToSegment(re, MultipleValue)
+	if seg != nil {
+		parser.haveLastSegment = true
+	}
+	return seg, err
+}
+
+// parseLiteral parses a segment with `Kind==Literal`, returning nil if not possible.
+func (parser *Parser) parseLiteral() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reLiteral, "^([a-zA-Z0-9_%]*)")
+	return parser.parseToSegment(re, Literal)
+}
+
+// parseToSegment is a helper function that creates a Segment of the specified kind if the next
+// characters in the parse stream match the expression re.
+func (parser *Parser) parseToSegment(re *regexp.Regexp, kind SegmentKind) (*Segment, error) {
+	match := parser.source.ConsumeRegex(re)
+	if len(match) == 0 {
+		return nil, nil
+	}
+	return &Segment{Kind: kind, Value: match}, nil
+}
+
+// parseFieldPath parses a field path, which is the "field" in a "{field=segments}" declaration.
+func (parser *Parser) parseFieldPath() string {
+	re := parser.GetRegexp(&parser.reFieldPath, "([a-zA-Z0-9_.]*)")
+	return parser.source.ConsumeRegex(re)
+}
+
+// parseVariable parses a variable spec, which is a sequence of field names, segments, and/or
+// placeholders, all enclosed by matching braces "{}".
+func (parser *Parser) parseVariable() (*Segment, error) {
+	if !parser.source.ConsumeIf('{') {
+		return nil, nil
+	}
+
+	fieldPath := parser.parseFieldPath()
+	if len(fieldPath) == 0 {
+		return nil, fmt.Errorf("no field path specified")
+	}
+
+	segment := &Segment{
+		Kind:  Variable,
+		Value: fieldPath,
+	}
+
+	if parser.source.ConsumeIf('=') {
+		var err error
+		segment.Subsegments, err = parser.parseSegments()
+		if err != nil {
+			return segment, err
+		}
+		if len(segment.Subsegments) == 0 {
+			return segment, fmt.Errorf("no path segments specified for field path %q", fieldPath)
+		}
+	} else {
+		segment.Subsegments = PathTemplate{&Segment{Kind: SingleValue}}
+	}
+
+	if !parser.source.ConsumeIf('}') {
+		return segment, fmt.Errorf("expected end-of-variable '}', got %q", parser.source.Str())
+	}
+
+	return segment, nil
+}
+
+// GetRegexp returns the memoized compiled regex corresponding to expr. This assumes the same re is always paired with the same expr.
+func (parser *Parser) GetRegexp(re **regexp.Regexp, expr string) *regexp.Regexp {
```

Comment:
I thought about it, but I really like the idea of each regexp being defined near where it's used (since each one is used in one place). Maybe that's too pedantic of me....